File: backend-node/src/modules/auth/auth.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../shared/prisma/prisma.service';
import { RSADecryptor } from './security/rsa.service';
import { PasswordService } from './security/password.service';
import { TokenService } from './jwt/jwt.service';
import { LoginDto, LoginResp } from './dto/login.dto';
import { UserInfo } from './models/user-info.model';
import { RouteItem } from './models/route-item.model';

/**
 * 认证与当前用户相关业务逻辑。
 */
@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly rsa: RSADecryptor,
    private readonly pwdService: PasswordService,
    private readonly tokenService: TokenService,
  ) {}

  /**
   * 登录流程，与 Go/Python 版保持一致。
   */
  async login(dto: LoginDto): Promise<LoginResp> {
    const authType = (dto.authType ?? '').trim().toUpperCase();
    if (authType && authType !== 'ACCOUNT') {
      throw new Error('暂不支持该认证方式');
    }
    if (!dto.clientId?.trim()) {
      throw new Error('客户端ID不能为空');
    }
    if (!dto.username?.trim()) {
      throw new Error('用户名不能为空');
    }
    if (!dto.password?.trim()) {
      throw new Error('密码不能为空');
    }

    let rawPassword: string;
    try {
      rawPassword = this.rsa.decryptBase64(dto.password);
    } catch {
      throw new Error('密码解密失败');
    }

    // 从 sys_user 查询用户
    const user = await this.prisma.sys_user.findFirst({
      where: { username: dto.username },
    });
    if (!user) {
      throw new Error('用户名或密码不正确');
    }

    const ok = this.pwdService.verify(rawPassword, user.password ?? '');
    if (!ok) {
      throw new Error('用户名或密码不正确');
    }
    if (user.status !== 1) {
      throw new Error('此账号已被禁用，如有疑问，请联系管理员');
    }

    const token = this.tokenService.generate(Number(user.id));
    return { token };
  }

  /**
   * 获取当前用户信息。
   */
  async getUserInfo(userId: number): Promise<UserInfo | null> {
    const user = await this.prisma.sys_user.findUnique({
      where: { id: BigInt(userId) },
    });
    if (!user) return null;

    // 角色
    const roles = await this.prisma.$queryRaw<
      { id: bigint; name: string; code: string; data_scope: number }[]
    >`
SELECT r.id, r.name, r.code, r.data_scope
FROM sys_role AS r
JOIN sys_user_role AS ur ON ur.role_id = r.id
WHERE ur.user_id = ${BigInt(userId)};
`;
    const roleCodes = roles.map((r) => r.code);

    // 权限
    const permsRows = await this.prisma.$queryRaw<{ permission: string }[]>`
SELECT DISTINCT m.permission
FROM sys_menu AS m
LEFT JOIN sys_role_menu AS rm ON rm.menu_id = m.id
LEFT JOIN sys_role AS r ON r.id = rm.role_id
LEFT JOIN sys_user_role AS ur ON ur.role_id = r.id
LEFT JOIN sys_user AS u ON u.id = ur.user_id
WHERE u.id = ${BigInt(userId)}
  AND m.status = 1
  AND m.permission IS NOT NULL;
`;
    const permissions = permsRows.map((p) => p.permission);

    // 部门名称
    let deptName = '';
    if (user.dept_id) {
      const dept = await this.prisma.sys_dept.findUnique({
        where: { id: user.dept_id },
        select: { name: true },
      });
      if (dept) {
        deptName = dept.name;
      }
    }

    const pwdResetTime = user.pwd_reset_time
      ? formatDateTime(user.pwd_reset_time)
      : '';
    const registrationDate = formatDateOnly(user.create_time);

    const opt = (v: string | null) => v ?? '';

    return {
      id: Number(user.id),
      username: user.username,
      nickname: user.nickname,
      gender: user.gender,
      email: opt(user.email),
      phone: opt(user.phone),
      avatar: opt(user.avatar),
      description: opt(user.description),
      pwdResetTime,
      pwdExpired: false,
      registrationDate,
      deptName,
      roles: roleCodes,
      permissions,
    };
  }

  /**
   * 获取当前用户路由树。
   */
  async getUserRoutes(userId: number): Promise<RouteItem[]> {
    const roles = await this.prisma.$queryRaw<
      { id: bigint; name: string; code: string; data_scope: number }[]
    >`
SELECT r.id, r.name, r.code, r.data_scope
FROM sys_role AS r
JOIN sys_user_role AS ur ON ur.role_id = r.id
WHERE ur.user_id = ${BigInt(userId)};
`;
    if (roles.length === 0) {
      return [];
    }
    const roleCodes = roles.map((r) => r.code);
    const roleIds = roles.map((r) => r.id);

    const menus = await this.prisma.$queryRaw<
      {
        id: bigint;
        parent_id: bigint;
        title: string;
        type: number;
        path: string | null;
        name: string | null;
        component: string | null;
        redirect: string | null;
        icon: string | null;
        is_external: boolean;
        is_cache: boolean;
        is_hidden: boolean;
        permission: string | null;
        sort: number;
        status: number;
      }[]
    >`
SELECT
  m.id,
  m.parent_id,
  m.title,
  m.type,
  m.path,
  m.name,
  m.component,
  m.redirect,
  m.icon,
  COALESCE(m.is_external, false) AS is_external,
  COALESCE(m.is_cache, false)    AS is_cache,
  COALESCE(m.is_hidden, false)   AS is_hidden,
  m.permission,
  COALESCE(m.sort, 0)            AS sort,
  m.status
FROM sys_menu AS m
JOIN sys_role_menu AS rm ON rm.menu_id = m.id
WHERE rm.role_id = ANY(${roleIds as any});
`;

    // 过滤掉按钮类型（3）
    const filtered = menus.filter((m) => Number(m.type) !== 3);
    if (filtered.length === 0) {
      return [];
    }

    // 排序：sort ↑，id ↑
    filtered.sort((a, b) => {
      if (a.sort === b.sort) {
        return Number(a.id - b.id);
      }
      return a.sort - b.sort;
    });

    const nodeMap = new Map<number, RouteItem>();
    for (const m of filtered) {
      const id = Number(m.id);
      nodeMap.set(id, {
        id,
        title: m.title,
        parentId: Number(m.parent_id),
        type: Number(m.type),
        path: m.path ?? '',
        name: m.name ?? '',
        component: m.component ?? '',
        redirect: m.redirect ?? '',
        icon: m.icon ?? '',
        isExternal: m.is_external,
        isHidden: m.is_hidden,
        isCache: m.is_cache,
        permission: m.permission ?? '',
        roles: roleCodes,
        sort: Number(m.sort),
        status: Number(m.status),
        children: [],
        activeMenu: '',
        alwaysShow: false,
        breadcrumb: true,
        showInTabs: true,
        affix: false,
      });
    }

    const roots: RouteItem[] = [];
    for (const item of nodeMap.values()) {
      if (item.parentId === 0) {
        roots.push(item);
        continue;
      }
      const parent = nodeMap.get(item.parentId);
      if (!parent) {
        roots.push(item);
        continue;
      }
      parent.children.push(item);
    }

    const sortChildren = (nodes: RouteItem[]) => {
      for (const node of nodes) {
        if (node.children.length > 0) {
          node.children.sort((a, b) => {
            if (a.sort === b.sort) {
              return a.id - b.id;
            }
            return a.sort - b.sort;
          });
          sortChildren(node.children);
        }
      }
    };
    sortChildren(roots);

    return roots;
  }
}

function pad(num: number): string {
  return num < 10 ? `0${num}` : `${num}`;
}

function formatDateOnly(date: Date): string {
  const y = date.getFullYear();
  const m = pad(date.getMonth() + 1);
  const d = pad(date.getDate());
  return `${y}-${m}-${d}`;
}

function formatDateTime(date: Date): string {
  const y = date.getFullYear();
  const m = pad(date.getMonth() + 1);
  const d = pad(date.getDate());
  const hh = pad(date.getHours());
  const mm = pad(date.getMinutes());
  const ss = pad(date.getSeconds());
  return `${y}-${m}-${d} ${hh}:${mm}:${ss}`;
}

File: backend-python/app/services/auth_service.py
from __future__ import annotations

from datetime import datetime
from typing import List, Dict, Any

from fastapi import Request

from ..api_response import fail, ok
from ..config import get_settings
from ..db import get_db_cursor
from ..models.auth import LoginRequest, LoginResponse, RouteItem, UserInfo
from ..security.jwt_token import TokenClaims, TokenService
from ..security.online_store import get_online_store
from ..security.password import PasswordVerifier
from ..security.rsa import RSADecryptor
from ..redis_client import redis_delete, redis_get


def get_token_service() -> TokenService:
    """构造 JWT 服务实例。"""
    s = get_settings()
    return TokenService(secret=s.jwt_secret, ttl_seconds=s.jwt_ttl_hours * 3600)


def get_rsa_decryptor() -> RSADecryptor:
    """构造 RSA 解密器实例。"""
    s = get_settings()
    return RSADecryptor.from_base64_key(s.rsa_private_key_b64)


def _is_login_captcha_enabled() -> bool:
    """读取 sys_option 中 LOGIN_CAPTCHA_ENABLED 配置，判断是否启用登录验证码。"""
    from ..db import get_db_cursor as _get_cursor

    with _get_cursor() as cur:
        cur.execute(
            """
SELECT COALESCE(value, default_value, '0') AS val
FROM sys_option
WHERE code = %s
LIMIT 1;
""",
            ("LOGIN_CAPTCHA_ENABLED",),
        )
        row = cur.fetchone()
    if not row:
        return False
    val = str(row["val"]).strip()
    return bool(val and val != "0")


def parse_token(auth_header: str | None, token_svc: TokenService) -> TokenClaims | None:
    """从 Authorization 头中解析用户信息。"""
    if not auth_header:
        return None
    return token_svc.parse(auth_header)


def login_service(
    req: LoginRequest,
    request: Request,
    token_svc: TokenService,
    decryptor: RSADecryptor,
):
    """账号密码登录服务实现，保持与原逻辑一致。"""
    auth_type = (req.authType or "").strip().upper()
    if auth_type and auth_type != "ACCOUNT":
        return fail("400", "暂不支持该认证方式")

    if not auth_type or auth_type == "ACCOUNT":
        if _is_login_captcha_enabled():
            if not (req.captcha or "").strip():
                return fail("400", "验证码不能为空")
            if not (req.uuid or "").strip():
                return fail("400", "验证码标识不能为空")
            captcha_key = f"CAPTCHA:{req.uuid.strip()}"
            stored = redis_get(captcha_key)
            if not stored:
                return fail("400", "验证码已过期")
            redis_delete(captcha_key)
            if (req.captcha or "").strip().lower() != stored.strip().lower():
                return fail("400", "验证码不正确")

    if not req.clientId.strip():
        return fail("400", "客户端ID不能为空")
    if not req.username.strip():
        return fail("400", "用户名不能为空")
    if not req.password.strip():
        return fail("400", "密码不能为空")

    try:
        raw_password = decryptor.decrypt_base64(req.password)
    except Exception:
        return fail("400", "密码解密失败")

    from ..db import get_db_cursor as _get_cursor

    with _get_cursor() as cur:
        cur.execute(
            """
SELECT
    id,
    username,
    nickname,
    password,
    gender,
    email,
    phone,
    avatar,
    description,
    status,
    is_system,
    pwd_reset_time,
    dept_id,
    create_user,
    create_time,
    update_user,
    update_time
FROM sys_user
WHERE username = %s
LIMIT 1;
""",
            (req.username,),
        )
        row = cur.fetchone()

    if not row:
        return fail("400", "用户名或密码不正确")

    encoded_pwd: str = row["password"]
    if not PasswordVerifier.verify(raw_password, encoded_pwd):
        return fail("400", "用户名或密码不正确")

    if row["status"] != 1:
        return fail("400", "此账号已被禁用，如有疑问，请联系管理员")

    user_id = int(row["id"])
    username = row["username"]
    nickname = row["nickname"]

    token = token_svc.generate(user_id)

    ip_header = ""
    user_agent = ""
    if request is not None:
        ip_header = request.headers.get("x-forwarded-for", "") or ""
        user_agent = request.headers.get("user-agent", "") or ""

    real_ip = ""
    if ip_header:
        real_ip = (ip_header.split(",")[0] or "").strip()
    if not real_ip and request is not None and request.client:
        real_ip = request.client.host or ""

    store = get_online_store()
    store.record_login(
        user_id=user_id,
        username=username,
        nickname=nickname,
        client_id=req.clientId,
        token=token,
        ip=real_ip,
        user_agent=user_agent,
    )

    resp = LoginResponse(token=token)
    return ok(resp.dict())


def logout_service(authorization: str | None, token_svc: TokenService):
    """登出服务实现。"""
    claims = parse_token(authorization, token_svc)
    if not claims:
        return fail("401", "未授权，请重新登录")

    authz = (authorization or "").strip()
    raw_token = authz
    lower = raw_token.lower()
    if lower.startswith("bearer "):
        raw_token = raw_token[7:].strip()

    if raw_token:
        store = get_online_store()
        store.remove_by_token(raw_token)

    return ok(claims.user_id)


def get_user_info_service(
    authorization: str | None,
    token_svc: TokenService,
):
    """获取当前登录用户信息服务实现。"""
    claims = parse_token(authorization, token_svc)
    if not claims:
        return fail("401", "未授权，请重新登录")

    user_id = claims.user_id
    with get_db_cursor() as cur:
        cur.execute(
            """
SELECT
    id,
    username,
    nickname,
    password,
    gender,
    email,
    phone,
    avatar,
    description,
    status,
    is_system,
    pwd_reset_time,
    dept_id,
    create_user,
    create_time,
    update_user,
    update_time
FROM sys_user
WHERE id = %s
LIMIT 1;
""",
            (user_id,),
        )
        u = cur.fetchone()
        if not u:
            return fail("401", "未授权，请重新登录")

        cur.execute(
            """
SELECT r.id, r.name, r.code, r.data_scope
FROM sys_role AS r
JOIN sys_user_role AS ur ON ur.role_id = r.id
WHERE ur.user_id = %s;
""",
            (user_id,),
        )
        roles = cur.fetchall()
        role_codes: List[str] = [r["code"] for r in roles]

        cur.execute(
            """
SELECT DISTINCT m.permission
FROM sys_menu AS m
LEFT JOIN sys_role_menu AS rm ON rm.menu_id = m.id
LEFT JOIN sys_role AS r ON r.id = rm.role_id
LEFT JOIN sys_user_role AS ur ON ur.role_id = r.id
LEFT JOIN sys_user AS u ON u.id = ur.user_id
WHERE u.id = %s
  AND m.status = 1
  AND m.permission IS NOT NULL;
""",
            (user_id,),
        )
        perms_rows = cur.fetchall()
        permissions: List[str] = [p["permission"] for p in perms_rows]

        dept_name = ""
        if u["dept_id"]:
            cur.execute(
                "SELECT name FROM sys_dept WHERE id = %s LIMIT 1;",
                (u["dept_id"],),
            )
            dept_row = cur.fetchone()
            if dept_row:
                dept_name = dept_row["name"]

    pwd_reset_time = ""
    if u["pwd_reset_time"]:
        pwd_reset_time = u["pwd_reset_time"].strftime("%Y-%m-%d %H:%M:%S")
    registration_date = u["create_time"].strftime("%Y-%m-%d")

    def _opt(v) -> str:
        return v or ""

    user_info = UserInfo(
        id=u["id"],
        username=u["username"],
        nickname=u["nickname"],
        gender=u["gender"],
        email=_opt(u["email"]),
        phone=_opt(u["phone"]),
        avatar=_opt(u["avatar"]),
        description=_opt(u["description"]),
        pwdResetTime=pwd_reset_time,
        pwdExpired=False,
        registrationDate=registration_date,
        deptName=dept_name,
        roles=role_codes,
        permissions=permissions,
    )
    return ok(user_info.dict())


def get_user_route_service(
    authorization: str | None,
    token_svc: TokenService,
):
    """获取当前用户的路由树结构服务实现。"""
    claims = parse_token(authorization, token_svc)
    if not claims:
        return fail("401", "未授权，请重新登录")
    user_id = claims.user_id

    with get_db_cursor() as cur:
        cur.execute(
            """
SELECT r.id, r.name, r.code, r.data_scope
FROM sys_role AS r
JOIN sys_user_role AS ur ON ur.role_id = r.id
WHERE ur.user_id = %s;
""",
            (user_id,),
        )
        roles = cur.fetchall()
        if not roles:
            return ok([])
        role_codes: List[str] = [r["code"] for r in roles]

        role_ids = [r["id"] for r in roles]
        cur.execute(
            """
SELECT
  m.id,
  m.parent_id,
  m.title,
  m.type,
  m.path,
  m.name,
  m.component,
  m.redirect,
  m.icon,
  COALESCE(m.is_external, false) AS is_external,
  COALESCE(m.is_cache, false)    AS is_cache,
  COALESCE(m.is_hidden, false)   AS is_hidden,
  m.permission,
  COALESCE(m.sort, 0)            AS sort,
  m.status
FROM sys_menu AS m
JOIN sys_role_menu AS rm ON rm.menu_id = m.id
WHERE rm.role_id = ANY(%s);
""",
            (role_ids,),
        )
        menu_rows = cur.fetchall()

    menus = [m for m in menu_rows if int(m["type"]) != 3]
    if not menus:
        return ok([])

    menus.sort(key=lambda m: (int(m["sort"]), int(m["id"])))

    node_map: dict[int, RouteItem] = {}
    for m in menus:
        node_map[int(m["id"])] = RouteItem(
            id=int(m["id"]),
            title=m["title"],
            parentId=int(m["parent_id"]),
            type=int(m["type"]),
            path=m.get("path") or "",
            name=m.get("name") or "",
            component=m.get("component") or "",
            redirect=m.get("redirect") or "",
            icon=m.get("icon") or "",
            isExternal=bool(m["is_external"]),
            isHidden=bool(m["is_hidden"]),
            isCache=bool(m["is_cache"]),
            permission=m.get("permission") or "",
            roles=role_codes,
            sort=int(m["sort"]),
            status=int(m["status"]),
            children=[],
        )

    roots: list[RouteItem] = []
    for item in node_map.values():
        if item.parentId == 0:
            roots.append(item)
            continue
        parent = node_map.get(item.parentId)
        if parent is None:
            roots.append(item)
            continue
        parent.children.append(item)

    def sort_children(nodes: list[RouteItem]) -> None:
        for node in nodes:
            if node.children:
                node.children.sort(key=lambda x: (x.sort, x.id))
                sort_children(node.children)

    sort_children(roots)

    return ok([r.dict() for r in roots])

